### ISSUE TITLE
mpc 0.34 requires mpd 0.21 and displays a warning

### ISSUE THREAD

**Issue by klundry** [mopidy/mopidy-mpd#47]
Everything appears to be still working normally as far as I can tell but mpc starting with v0.34 prints this warning with all output and I don't see any way to disable it. Also, the warning text is somehow immune to tail/grep. It will print no matter what. Quite annoying.

`warning: MPD 0.21 required`

**Comment by girst** [comment]
On Tue, Dec 07, 2021 at 01:46:39PM -0800, klundry wrote:
>Everything appears to be still working normally as far as I can tell 
>but mpc starting with v0.34 prints this warning with all output and I 
>
>`warning: MPD 0.21 required`

mopidy-mpd advertises 0.19 compatibility (c.f.  
https://github.com/mopidy/mopidy-mpd/blob/1ab15a0c6b34a6cfb32674fa0b6f867decf6d8cf/mopidy_mpd/protocol/__init__.py#L24). 
the warning was introduced a week ago in 
https://github.com/MusicPlayerDaemon/mpc/commit/ce1a0047c6ca0c4afcc70495447bdd18dbce7ccb

the checks are gating send_tag_types_for_format() and 
mpd_search_add_expression(). tagtypes is not supported, looking at 
mopidy-mpd's readme; i'm not sure about search expressions, but probably 
not supported either.

maybe someone has the time to implement those, otherwise i don't think 
they'd be a showstopper (you can still use mpc, just that you might get 
an error when sending tagtypes or using search expressions). 

>don't see any way to disable it. Also, the warning text is somehow 
>immune to tail/grep. It will print no matter what. Quite annoying.

it is probably printed to stderr. you can add 2>/dev/null to the mpc 
call to hide all error and warning messages (but possibly even those 
that you care about).

**Comment by klundry** [comment]
Thanks for the tip about stderr, I will try that. I'm only using pretty basic functions of mpc like playing playlists, querying the current song in some scripts. If I can just get rid of the warning text that will be good enough for me until the newer things get implemented.

**Comment by jose1711** [comment]
FWIW even though `mpc` works with warning, `ncmpc` client breaks.

**Comment by JoshMock** [comment]
Experiencing the same issue with `ncmpc` as of this morning. Error on startup:

```
Error: MPD version 0.19.0 is too old (0.21.0 needed)
```

System details:
- Arch Linux `5.18.9-arch1-1`
- Mopidy versions:
    ```
    mopidy 3.3.0-1
    mopidy-beets 4.0.1-2
    mopidy-mpd 3.3.0-1
    mopidy-mpris 3.0.3-2
    mopidy-scrobbler 2.0.1-2
    mopidy-youtube v3.5.0-2
    python-mopidy-tidal 0.2.7-1
    ```

Switched to the Iris web-based client for the time being, but I use `mpc` to automate a lot of local shortcuts, so if/when that stops working I'm out of luck.

I've not contributed to Mopidy or mopidy-mpd before, but would be happy to contribute a fix if someone can coach me through getting started.

**Comment by girst** [comment]
On Tue, Jul 05, 2022 at 09:10:20AM -0700, Josh Mock wrote:
>Experiencing the same issue with `ncmpc` as of this morning. Error on startup:
>
>```
>Error: MPD version 0.19.0 is too old (0.21.0 needed)

this is still waiting on our support for search expressions (afaik, 
that's the only 0.21 stuff we're still missing). I've started working on 
that in PR#52, but i got distracted with other stuff. if someone wants 
to pick that up, go for it! otherwise, it'll take me some time to finish 
that.

the problem of right now is that the parser is way too complicated and 
mopidy's api (and backends themselves) can't support the general form of 
these expressions. it'd be best if we could at least handle simple 
cases, and just fudge the results when encountering the other cases.

as long as it continues to be just a warning (and not an error that 
exits the program), i'm inclined to say we just live with it.

**Comment by JoshMock** [comment]
> as long as it continues to be just a warning (and not an error that exits the program), i'm inclined to say we just live with it.

In my case, `ncmpc` has fully stopped working since this came up, so may be it requiring the use of a non-MPD client to interact with Mopidy until fixed.

**Comment by jodal** [comment]
> *This was generated by AI during triage.*

The warning comes because Mopidy-MPD implements protocol version 0.19. The fix is support for a newer protocol version, which #22 tracks. Closing as a duplicate of #22.
